services/src/main/java/com/chatbot/services/ChatServiceController.java

```diff
@@ -0,0 +1,185 @@
+        chatServiceRequestBuilder.setRequestType(RequestType.MESSAGE);
+        chatServiceRequestBuilder = parseHangoutsCardClick(chatServiceRequestBuilder, event);
+        break;
+        // throw new IllegalArgumentException("Nothing much");
+      default:
+        throw new IllegalArgumentException("The request has no event type");
+    }
+    chatServiceRequestBuilder = parseHangoutsSender(chatServiceRequestBuilder, event);
+    return chatServiceRequestBuilder.build();
+  }
+
+  private ChatServiceRequest.Builder parseHangoutsUserMessage(
+      final ChatServiceRequest.Builder chatServiceRequestBuilder, final JsonNode event) {
+    final ChatServiceRequest.UserMessage.Builder userMessageBuilder =
+        ChatServiceRequest.UserMessage.newBuilder();
+    if (event.at("/message").has("attachment")) {
+      if (event.at("/message").has("argumentText")) {
+        userMessageBuilder.setText(event.at("/message/argumentText").asText());
+      }
+      final Iterator<JsonNode> attachmentIterator = event.at("/message/attachment").elements();
+      while (attachmentIterator.hasNext()) {
+        final JsonNode attachment = (JsonNode) attachmentIterator.next();
+        final ChatServiceRequest.Attachment.Builder attachmentBuilder =
+            ChatServiceRequest.Attachment.newBuilder();
+        attachmentBuilder.setLink(attachment.at("/downloadUri").asText());
+        switch (attachment.at("/contentType").asText()) {
+          case "image/png":
+            attachmentBuilder.setMimeType(MimeType.PNG);
+            break;
+          case "image/jpeg":
+            attachmentBuilder.setMimeType(MimeType.JPEG);
+            break;
+          default:
+            attachmentBuilder.setMimeType(MimeType.UNKNOWN_MIME_TYPE);
+        }
+        userMessageBuilder.addAttachments(attachmentBuilder);
+      }
+    } else {
+      userMessageBuilder.setText(event.at("/message/text").asText());
+    }
+    chatServiceRequestBuilder.setUserMessage(userMessageBuilder);
+    return chatServiceRequestBuilder;
+  }
+
+  private ChatServiceRequest.Builder parseHangoutsCardClick(
+      final ChatServiceRequest.Builder chatServiceRequestBuilder, final JsonNode event) {
+    final ChatServiceRequest.UserMessage.Builder userMessageBuilder =
+        ChatServiceRequest.UserMessage.newBuilder();
+    userMessageBuilder.setText(event.at("/action/parameters/0/value").asText());
+    chatServiceRequestBuilder.setUserMessage(userMessageBuilder);
+    return chatServiceRequestBuilder;
+  }
+
+  private ChatServiceRequest.Builder parseHangoutsSender(
+        final ChatServiceRequest.Builder chatServiceRequestBuilder, final JsonNode event) {
+    final ChatServiceRequest.Sender.Builder senderBuilder = ChatServiceRequest.Sender.newBuilder()
+        .setDisplayName(event.at("/user/displayName").asText())
+        .setChatClientGeneratedId(event.at("/space/name").asText().substring(7))
+        .setUserId(event.at("/user/name").asText().substring(6));
```

Comment:
Can we use constants instead of 6,7 hard to understand stuff.

Reply:
updated